pkg/engine/datasource/graphql_datasource/graphql_subscription_client.go

```diff
@@ -48,135 +110,282 @@ func (c *WebSocketGraphQLSubscriptionClient) Subscribe(ctx context.Context, opti
+	next    chan<- []byte
+}
+
+func (h *connectionHandler) startBlocking(sub subscription) {
+	readCtx, cancel := context.WithCancel(h.ctx)
 	defer func() {
-		close(next)
-		_ = conn.Write(ctx, websocket.MessageText, []byte(fmt.Sprintf(stopMessage, "1")))
-		_ = conn.Close(websocket.StatusNormalClosure, "")
+		h.unsubscribeAllAndCloseConn()
+		cancel()
 	}()
-	subscriptionLifecycle := ctx.Done()
-	resolverLifecycle := c.ctx.Done()
+	h.subscribe(sub)
+	dataCh := make(chan []byte)
+	go h.readBlocking(readCtx, dataCh)
 	for {
-		select {
-		case <-subscriptionLifecycle:
+		if h.ctx.Err() != nil {
 			return
-		case <-resolverLifecycle:
+		}
+		hasActiveSubscriptions := h.checkActiveSubscriptions()
+		if !hasActiveSubscriptions {
 			return
-		default:
-			msgType, data, err := conn.Read(ctx)
-			if err != nil {
-				continue
-			}
-			if msgType != websocket.MessageText {
-				continue
-			}
+		}
+		select {
+		case <-time.After(h.readTimeout):
+			continue
+		case sub = <-h.subscribeCh:
+			h.subscribe(sub)
+		case data := <-dataCh:
 			messageType, err := jsonparser.GetString(data, "type")
 			if err != nil {
 				continue
 			}
 			switch messageType {
 			case "data":
-				id, err := jsonparser.GetString(data, "id")
-				if err != nil {
-					continue
-				}
-				if id != "1" {
-					continue
-				}
-				payload, _, _, err := jsonparser.Get(data, "payload")
-				if err != nil {
-					continue
-				}
-				select {
-				case next <- payload:
-					continue
-				case <-subscriptionLifecycle:
-					return
-				}
+				h.handleMessageTypeData(data)
 			case "complete":
-				return
+				h.handleMessageTypeComplete(data)
 			case "connection_error":
-				next <- []byte(`{"errors":[{"message":"connection error"}]}`)
+				h.handleMessageTypeConnectionError()
 				return
 			case "error":
-				value, valueType, _, err := jsonparser.Get(data, "payload")
-				if err != nil {
-					next <- []byte(`{"errors":[{"message":"internal error"}]}`)
-					return
-				}
-				switch valueType {
-				case jsonparser.Array:
-					response := []byte(`{}`)
-					response, err = jsonparser.Set(response, value, "errors")
-					if err != nil {
-						next <- []byte(`{"errors":[{"message":"internal error"}]}`)
-						return
-					}
-					next <- response
-				case jsonparser.Object:
-					response := []byte(`{"errors":[]}`)
-					response, err = jsonparser.Set(response, value, "errors", "[0]")
-					if err != nil {
-						next <- []byte(`{"errors":[{"message":"internal error"}]}`)
-						return
-					}
-					next <- response
-				default:
-					next <- []byte(`{"errors":[{"message":"internal error"}]}`)
-				}
-				return
+				h.handleMessageTypeError(data)
+				continue
 			default:
 				continue
 			}
 		}
 	}
 }
+
+func (h *connectionHandler) readBlocking(ctx context.Context, dataCh chan []byte) {
+	for {
+		msgType, data, err := h.conn.Read(ctx)
+		if err != nil {
+			continue
+		}
+		if msgType != websocket.MessageText {
+			continue
+		}
+		dataCh <- data
+	}
```

Comment:
It could be disastrous if there is no way to exit this loop(like failing to read forever when h.conn is closed), we could use the context:
```go
        done := ctx.Done()
	for {
		select {
		case <-done:
			return
		default:
			msgType, data, err := h.conn.Read(ctx)
			if err != nil {
				continue
			}
			if msgType != websocket.MessageText {
				continue
			}
			dataCh <- data
		}
	}
```